src/server.c

```diff
@@ -2911,46 +2911,64 @@ void resetErrorTableStats(void) {
 
 /* ========================== Redis OP Array API ============================ */
 
-void redisOpArrayInit(redisOpArray *oa) {
-    oa->ops = NULL;
+void redisOpArrayReset(redisOpArray *oa) {
     oa->numops = 0;
-    oa->capacity = 0;
+    oa->len = 0;
 }
 
 int redisOpArrayAppend(redisOpArray *oa, int dbid, robj **argv, int argc, int target) {
     redisOp *op;
     int prev_capacity = oa->capacity;
 
-    if (oa->numops == 0) {
+    if (oa->len == 0) {
         oa->capacity = 16;
-    } else if (oa->numops >= oa->capacity) {
+    } else if (oa->len >= oa->capacity) {
         oa->capacity *= 2;
     }
 
     if (prev_capacity != oa->capacity)
         oa->ops = zrealloc(oa->ops,sizeof(redisOp)*oa->capacity);
-    op = oa->ops+oa->numops;
+    op = oa->ops+oa->len;
+    op->dbid = dbid;
+    op->argv = argv;
+    op->argc = argc;
+    op->target = target;
+    oa->len++;
+    if (op->target) {
+        oa->numops++;
+    }
+    return oa->len;
+}
+
+int redisOpArrayAppendPlaceholder(redisOpArray *oa) {
+    return redisOpArrayAppend(oa, 0, NULL, 0, 0) - 1;
+}
+
+void redisOpArraySet(redisOpArray *oa, int dbid, robj **argv, int argc, int target, int index) {
+    redisOp *op = oa->ops+index;
     op->dbid = dbid;
     op->argv = argv;
     op->argc = argc;
     op->target = target;
     oa->numops++;
-    return oa->numops;
 }
 
 void redisOpArrayFree(redisOpArray *oa) {
-    while(oa->numops) {
+    while(oa->len) {
         int j;
         redisOp *op;
 
-        oa->numops--;
-        op = oa->ops+oa->numops;
-        for (j = 0; j < op->argc; j++)
+        oa->len--;
+        op = oa->ops+oa->len;
+        if (!op->target) {
+            continue;
+        }
+        for (j = 0; j < op->argc; j++) {
             decrRefCount(op->argv[j]);
+        }
         zfree(op->argv);
     }
-    zfree(oa->ops);
-    redisOpArrayInit(oa);
+    redisOpArrayReset(oa);
```

Comment:
what about `zfree(oa->ops)`?

Reply:
We reuse the array so we will not have to allocated it each time.